src/permissions/NativeTokenRollingSpendLimit/NativeTokenRollingSpendLimitPermission.sol

```diff
@@ -66,18 +74,27 @@ contract NativeTokenRollingSpendLimitPermission is IPermissionContract {
         ICoinbaseSmartWallet.Call[] memory calls = abi.decode(userOp.callData[4:], (ICoinbaseSmartWallet.Call[]));
         uint256 callsLen = calls.length;
         uint256 spendValue = 0;
-        // if no paymaster, set initial spendValue as requiredPrefund
-        /// @dev no accounting done for the refund step so ETH spend is overestimated slightly
-        if (userOp.paymasterAndData.length == 0) {
+
+        // increment spendValue if gas cost beared by the user
+        /// @dev No accounting done for the refund step so native token spend is overestimated slightly
+        /// @dev MagicSpend only allows withdrawing native token when used as a paymaster
+        /// @dev Rely on Coinbase Cosigner to prevent use of paymasters that spend user assets not tracked here
+        if (
+            userOp.paymasterAndData.length == 0 || address(bytes20(userOp.paymasterAndData[:20])) == MAGIC_SPEND_ADDRESS
+        ) {
             spendValue += UserOperationUtils.getRequiredPrefund(userOp);
```

Comment:
Might be worth commenting so we remember that this will probably overdebit by 3x actual gas used.